packages/fcl/src/wallet-provider-spec/draft-v3.md

```diff
@@ -0,0 +1,569 @@
+|----------------|--------|-------|
+| HTTP/POST      |   ⛔   |   ✅   |
+| IFRAME/RPC     |   ✅   |   ⛔   |
+| POP/RPC        |   ✅   |   ⛔   |
+| TAB/RPC        |   ✅   |   ⛔   |
+
+It's important to note that regardless of the method of communication, the data that is sent back and forth between the parties involved is the same.
+
+### IFRAME/RPC
+
+`IFRAME/RPC` is the easiest to explain, so we will start with it:
+
+- An iframe is rendered (comes from the `endpoint` in the service).
+- The rendered frames says its ready `WalletUtils.sendMsgToFCL("FCL:VIEW:READY")`.
+- FCL will send the data to be dealt with: `WalletUtils.sendMsgToFCL("FCL:VIEW:READY:RESPONSE", {...body, service: {params, data} })`
+  - Where `body` is the stuff you care about, `params` and `data` are additional information you can provide in the service object.
+- The wallet sends back an Approved or Declined post message. (It will be a `f_type: "PollingResponse"`, which we will get to in a bit)
+  - If it's approved, the polling responses data field will need to be what FCL is expecting.
+  - If it's declined, the polling responses reason field should say why it was declined.
+
+![IFRAME/RPC Diagram](https://raw.githubusercontent.com/onflow/flow-js-sdk/master/packages/fcl/assets/service-method-diagrams/iframe-rpc.png)
+
+### POP/RPC
+
+`POP/RPC` works in an almost entirely similar way to `IFRAME/RPC`, except instead of rendering the `method` in an iframe, we render it in a popup. The same communication protocol between the rendered view and FCL applies:
+
+- A popup is rendered (comes from `endpoint` in the service).
+- The rendered popup says its ready `WalletUtils.sendMsgToFCL("FCL:VIEW:READY")`.
+- FCL will send the data to be dealt with `WalletUtils.sendMsgToFCL("FCL:VIEW:READY:RESPONSE", { ...body, service: {params, data} })`
+  - Where `body` is the stuff you care about, `params` and `data` are additional information you can provide in the service object.
+- The wallet sends back an Approved or Declined post message (It will be a `f_type: "PollingResponse"`, we will get to that in a bit)
+  - If it's approved, the polling responses data field will need to be what FCL is expecting.
+  - If it's declined, the polling responses reason field should say why it was declined.
+
+![POP/RPC Diagram](https://raw.githubusercontent.com/onflow/flow-js-sdk/master/packages/fcl/assets/service-method-diagrams/pop-rpc.png)
+
+### TAB/RPC
+
+`TAB/RPC` works in an almost entirely similar way to `IFRAME/RPC` and `POP/RPC`, except instead of rendering the `method` in an iframe or a popup, we render it in a new tab. The same communication protocol between the rendered view and FCL applies:
+
+- A new tab is rendered (comes from `endpoint` in the service).
+- The rendered tab says its ready `WalletUtils.sendMsgToFCL("FCL:VIEW:READY")`.
+- FCL will send the data to be dealt with `WalletUtils.sendMsgToFCL("FCL:VIEW:READY:RESPONSE", { ...body, service: {params, data} })`
```

Comment:
`sendMsgToFCL` -> `onMsgFromFCL`

Probably makes more sense to state this in terms of `onMsgFromFCL`